### PR TITLE
[clang] Add `intrin0.h` header to mimic `intrin0.h` used by MSVC STL for clang-cl

### LINKED ISSUE #53520: clang-cl can not support function targets
The immintrin.h headers has a bug where it does not include sub-arch intrinsics headers if _MSC_VER is defined unless the subarch target is current.

This is inconsistent with MSVC which always defines intrinsics regardless of active arch, and is also inconsistent with normal-clang (gcc-style) which defines them for use with subarch function targets. The result is there is no way of using intrinsics for subarch targets with clang-cl, since neither gcc nor msvc style works.

This has forced us to disable many optimizations in Qt with clang-cl, see https://bugreports.qt.io/browse/QTBUG-88081, https://bugreports.qt.io/browse/QTBUG-88434 and https://bugreports.qt.io/browse/QTBUG-98253

I suggest at least allowing gcc style, and the intrinsics working in target attributed functions, if you can't support the MSVC style.

### PR DESCRIPTION
**Pull request by MaxEW707**
Fixes https://github.com/llvm/llvm-project/issues/53520.

#### Description ####

Provide `intrin0.h` to be the minimal set of intrinsics that the MSVC STL requires.
The `intrin0.h` header matches the latest header provided by MSVC 1939 which does include some extra intrinsics that the MSVC STL does not use.

Inside `BuiltinHeaders.def` I kept the header description as `intrin.h`. If you want me to change those to `intrin0.h` for the moved intrinsics let me know.

This should now allow `immintrin.h` to be used with function targets for runtime cpu detection of simd instruction sets without worrying about the compile-time overhead from MSVC STL including `intrin.h` on clang.

I still need to figure out how to best update MSVC STL to detect for the presence of `intrin0.h` from clang and to use this header over `intrin.h`.

#### Testing ####

Built clang locally and ran the test suite. I still need to do a pass over the existing unit tests for the ms intrinsics to make sure there aren't any gaps. Wanted to get this PR up for discussion first.

Modified latest MSVC STL from github to point to `intrin0.h` for clang.

Wrote some test files that included MSVC STL headers that rely on intrinsics such as `atomic`, `bit` and `vector`. Built the unit tests against x86, arm, aarch64, and x64.

#### Benchmarks ####

The following include times are based on the x64 target with the modified headers in this PR.
These timings were done by using `clang-cl.exe -ftime-trace` and taking the wall time for parsing `intrin.h` and `intrin0.h`.

`intrin.h` takes ~897ms to parse.
`intrin0.h` takes ~1ms to parse.

If there is anything required or a different approach is preferred let me know. I would very much like to move this over the finish line so we can use function targets with clang-cl.

### CODE REVIEW

**Comment by nico** [comment]
Is `intrin0.h` a header that ships with MSVC's compiler, or with MS's STL? If the latter, shouldn't everything just work already?

**Comment by nico** [comment]
Looks like it belongs to MSVC: https://github.com/microsoft/STL/blob/a8888806c6960f1687590ffd4244794c753aa819/stl/inc/yvals_core.h#L2040

So I think the right order here is:

1. Add intrin0.h header
2. Wait for a release
3. Change MSSTL to include that in clang builds as well
4. Wait for a release of that
5. Make intrin.h the chonky header

If we land this as-is, it'll tank build time on Windows.

**Comment by nico** [comment]
In other words, please undo the changes to `clang/lib/Headers/immintrin.h` for now.

**Comment by rnk** [comment]
> If we land this as-is, it'll tank build time on Windows.

While this is true, I don't think it's the right tradeoff for us to leave Intel intrinsics inaccessible for users who don't want to enable new microarchitectural features globally with command line flags. You may recall there are ODR issues, where enabling AVX globally results in AVX instructions being generated in inline functions, which then prevail at link time, and crash at runtime on chips where AVX is not available. We are currently prioritizing compile time above those needs, and that doesn't seem like the right tradeoff.

What I'd like to see is a pull request sent to https://github.com/microsoft/stl with some agreement about how to structure the ifdefs so we can use intrin0.h when it is available. We probably can't use `__has_include(<intrin0.h>)` because MSVC ships its own copy of intrin0.h, and they can't be distinguished. Perhaps we should come up with our own name for `intrin0.h`, so we can feature-detect it. I suggest `intrin_msstl.h` or `intrin_minimal.h` or something.

Then, when we land this PR, the net effect is that compile time with MSVC slows down until the next release of the Microsoft STL, but users have access to Intel intrinsics as soon as possible. Does that seem reasonable?

**Comment by MaxEW707** [comment]
> Is intrin0.h a header that ships with MSVC's compiler, or with MS's STL? If the latter, shouldn't everything just work already?

It ships with MSVC.

`intrin0.h` from MSVC doesn't work currently due to some declarations not matching declarations in clang's own `intrin.h`.
For example `_addcary_u64`. Which I now realize I am missing since that is inside `adxintrin.h` which I mistakenly assumed only included adx instriniscs and not adc intrinsics as well.

> What I'd like to see is a pull request sent to https://github.com/microsoft/stl with some agreement about how to structure the ifdefs so we can use intrin0.h when it is available.

Sounds good I'll do that.

> We probably can't use __has_include(<intrin0.h>) because MSVC ships its own copy of intrin0.h, and they can't be distinguished.

I was thinking I could use `__has_include_next`. If `_has_include_next(<intrin0.h>)` is true then we know clang is providing its own `intrin0.h`.

Another option I had was to preemptively get a PR up to MSVC STL with a `__clang_major__` check for clang 18 assuming this PR gets in before the release window.

A custom name like `intrin_msstl.h` works for me as well.

**Comment by rnk** [comment]
> > What I'd like to see is a pull request sent to https://github.com/microsoft/stl with some agreement about how to structure the ifdefs so we can use intrin0.h when it is available.
> 
> Sounds good I'll do that.

Thanks!

> I was thinking I could use `__has_include_next`. If `__has_include_next(<intrin0.h>)` is true then we know clang is providing its own `intrin0.h`.
> 
> Another option I had was to preemptively get a PR up to MSVC STL with a `__clang_major__` check for clang 18 assuming this PR gets in before the release window.
> 
> A custom name like `intrin_msstl.h` works for me as well.

I think we can be flexible, and the important thing is that our solution should work for them.

**Comment by MaxEW707** [comment]
MSVC STL requires the x64 adc intel intrinsics. I moved those to a separate file `adcintrin.h` that can be included from `immintrin.h` and `intrin0.h` for x64.

I also made a table [here ](https://gist.github.com/MaxEW707/2f4bf26801bc1b6b088aa7a2fadba526) of all the intrinsics available with MSVC's `intrin0.h`, which intrinsics are used by MSVC STL, which are used by MSVC STL when compiled under clang and which intrinsics are provided as builtins from clang.

You will notice that the `_ReadWriteBarrier` compiler barrier isn't supported on arm by clang.  This [godbolt](https://godbolt.org/z/xEv3oEbK6) appears to corroborate that finding. I am not too worried about this since I believe Microsoft is dropping 32-bit ARM in a future Windows 11 update if my memory is correct.

The last table at the end shows the intrinsics that aren't implemented in clang but MSVC STL has workarounds using the clang intrinsics instead. We don't have to worry about supporting those.

**Comment by RKSimon** [comment]
@MaxEW707 If you pull out the adcintrin.h change into a separate PR we can fast track it as it should be done anyhow

**Comment by MaxEW707** [comment]
> @MaxEW707 If you pull out the adcintrin.h change into a separate PR we can fast track it as it should be done anyhow

Done :). Here is the link since I can't add reviewers https://github.com/llvm/llvm-project/pull/75992.

File: clang/lib/Headers/adcintrin.h
Comment (phoebewang) [inline clang/lib/Headers/adcintrin.h:1]: adx -> adc

**Comment by AaronBallman** [comment]
> If we land this as-is, it'll tank build time on Windows.

I would like some measurements so we can compare build times on Windows.

**Comment by MaxEW707** [comment]
> I would like some measurements so we can compare build times on Windows.

I took some benchmarks with `-ftime-trace` on the parse times with and without this change.
Pretty much all the big hitters, string/vector/map/algorithm, includes `<xmemory>` which includes `<limits>` which includes `<intrin[0].h>`.
The test file used looked as follows to simulate some common stl includes.
```
#include <map>
#include <vector>
#include <memory>
```

clang-cl 16 frontend took ~190ms to parse those 3 headers. `intrin.h` took ~32ms to parse.

clang-cl built with this PR frontend took ~1368ms to parse. `intrin.h` took ~969ms to parse. Most of that time is from `<x86intrin.h>` as expected. `intrin0.h` took ~2ms to parse.

It is known that `immintrin.h` and `x86intrin.h` being the everything header definitely makes it an **expensive** include.
I don't need to compile any project to know that include time difference will be felt by any code that uses STL.

MSVC STL ships with each compiler upgrade. VS has a separate dir for each installed msvc, `VC\Tools\MSVC\14.38.33130` for MSVC 1938, and each installed compiler has a specific version of msvc stl shipped with it under, `VC\Tools\MSVC\[version]\include`.
Only one version of clang-cl is shipped with VS at a time and it will always point to the latest `VCToolsDir` via the clang driver unless clang is passed an explicit dir on the command line by the user.
For example VS2022 17.7 upgraded from clang-cl 15 to clang-cl 16. From that point on-wards only clang-cl 16 exists on disk in your VS install.
That seems to correspond to what was said here, https://github.com/microsoft/STL/pull/4282#discussion_r1441191186, every shipped version of MSVC STL supports only the latest compilers provided by VS.

We have some options. I am fine with whatever the consensus is as long as I can move this over the finish line.
1. Users who upgrade clang-cl before MSVC STL officially supports the new version of clang-cl will suffer slower builds.
    I would classify this as an unsupported config since once VS ships with support for the new clang-cl, intentionally building against an older `VCToolsDir` will also incur the include overhead. Up to you guys if view this as an unsupported config.
2. Add a config define that users can define to enable the old behaviour in the interim. Once MSVC STL supports the new version of clang-cl we can remove this config define in the next release.
3.  Release clang with `intrin0.h`. Get a change into MSVC STL. Wait for MSVC STL release. Release another clang with `intrin.h` changes. I am not well versed on what changes clang allows in patch or minor releases but this feels like it would have a long lag time before clang-cl can become usable for us.

**Comment by rnk** [comment]
> > I would like some measurements so we can compare build times on Windows.
> 
> I took some benchmarks with `-ftime-trace` on the parse times with and without this change. 
> ...
> clang-cl built with this PR frontend took ~1368ms to parse. `intrin.h` took ~969ms to parse. Most of that time is from `<x86intrin.h>` as expected. `intrin0.h` took ~2ms to parse.

Thank you for remeasuring! Roughly speaking, anyone using a clang from between the time that this PR lands and they update the MSVC STL will experience a 1 second slowdown in every compilation that didn't already use `immintrin.h`.

> We have some options. I am fine with whatever the consensus is as long as I can move this over the finish line.
> 
> 1. Users who upgrade clang-cl before MSVC STL officially supports the new version of clang-cl will suffer slower builds.
>    I would classify this as an unsupported config since once VS ships with support for the new clang-cl, intentionally building against an older `VCToolsDir` will also incur the include overhead. Up to you guys if view this as an unsupported config.
> 2. Add a config define that users can define to enable the old behaviour in the interim. Once MSVC STL supports the new version of clang-cl we can remove this config define in the next release.
> 3. Release clang with `intrin0.h`. Get a change into MSVC STL. Wait for MSVC STL release. Release another clang with `intrin.h` changes. I am not well versed on what changes clang allows in patch or minor releases but this feels like it would have a long lag time before clang-cl can become usable for us.

I think I favor 2. Users (Chromium included) use recent versions of Clang that are not vendored by Microsoft. Users can also use the LLVM OSS releases. I think most users probably won't notice the compile time regression, and it will silently go away the next time the update MSVC, but if they do and they care (and this should be in the release notes), it would be nice to have a macro to get the old behavior. There may also be users permanently targeting old MSVC STL versions and they will never get the update to use `intrin0.h`, but I believe fewer people do this since MSVC stabilized the standard library ABI.

It would be nice to factor out the `!defined(__SCE__) || __has_feature(modules)` condition into something more descriptive like `__CLANG_ENABLE_ALL_INTEL_INTRIN`, to clean up the conditional includes regardless of whether we add this opt-out flag.

**Comment by MaxEW707** [comment]
> Users (Chromium included) use recent versions of Clang that are not vendored by Microsoft. Users can also use the LLVM OSS releases. I think most users probably won't notice the compile time regression, and it will silently go away the next time the update MSVC, but if they do and they care (and this should be in the release notes), it would be nice to have a macro to get the old behavior. 

I thought of another solution this morning that can work. Looks like we already have precedent with redefining macros in clang's interposing MSVC headers. `vadefs.h` for example.

MSVC STL would like to replace [_STL_INTRIN_HEADER](https://github.com/microsoft/STL/blob/main/stl/inc/yvals_core.h#L2038) macro with just `#include <intrin0.h>` when they support the version of clang shipped with `intrin0.h`.

We could provide our own `yvals_core.h` header like so:
```
#include_next <yvals_core.h>

#if defined(_MSC_VER_) && defined(_STL_INTRIN_HEADER)
#undef _STL_INTRIN_HEADER
#define _STL_INTRIN_HEADER #include <intrin0.h>
#endif
```

This should also fix users upgrading clang while still building against an older MSVC STL. Once MSVC STL moves forward we can probably remove this interposing header sometime in the future.

**Comment by MaxEW707** [comment]
> #include <memory>
>
>clang-cl 16 frontend took ~190ms to parse those 3 headers. intrin.h took ~32ms to parse.
>
>clang-cl built with this PR frontend took ~1368ms to parse. intrin.h took ~969ms to parse. Most of that time is from <x86intrin.h> as expected. intrin0.h took ~2ms to parse.

Did this same test with the most recent change that included the `yvals_core.h` interposing header.
The parse time was around ~170ms since `intrin0.h` is still cheaper than the old `intrin.h`.

This was tested with the MSVC STL shipped with MSVC 1938.

**Comment by MaxEW707** [comment]
Ping

Wondering if there is anything else required of me to progress this PR forward. Let me know :).

**Comment by nico** [comment]
Wow, that clang/lib/Headers/yvals_core.h hacks is gross, I love it :D

lgtm, and good to land provided you do a quick re-check to confirm that this doesn't tank compile times now with that part included. Thanks!

**Review by rnk — APPROVED** [review]
I think there are risks to shadowing `yvals_core.h`, but @CaseyCarter is at least aware of it [here](https://github.com/microsoft/STL/issues/3634#issuecomment-1904956652).

Once Clang ships its own intrin0.h header, the MSVC STL can adjust their yvals_core.h code to use it, and then we can delete this shadow header. This gives us the best outcome: Fast compiles today, target attribute support in MSVC-mode, and minimal tech debt in the long run. Sounds good to me, thanks for finding a solution!

File: clang/lib/Headers/yvals_core.h
Comment (MaskRay) [inline clang/lib/Headers/yvals_core.h:10]: Newer lib/Headers changes prefer `//`-style comments. Do we need `/* */`?

**Comment by MaskRay** [comment]
I am not familiar with the MSVC ecosystem, but the change looks reasonable. Thanks!

**Comment by MaxEW707** [comment]
I will need someone to commit on my behalf since I do not have write access.

> lgtm, and good to land provided you do a quick re-check to confirm that this doesn't tank compile times now with that part included. Thanks!

The include times below are done with the following source file using a locally release build of clang, `-ftime-trace`, and msvc stl shipped with MSVC 1939.
```
#include <map>
#include <vector>
#include <memory>
```

clang-cl.exe without `yvals_core.h` shadowing takes ~1,344 ms in the frontend. `intrin.h` took ~955 ms to parse.

clang-cl.exe with `yvals_core.h` shadowing takes ~395 ms in the frontend. `intrin0.h` took ~1.4 ms to parse.

File: clang/lib/Headers/yvals_core.h
Comment (MaxEW707) [reply clang/lib/Headers/yvals_core.h:10]: We do not need `/* */`.

I was just copying from the other files in this folder for the copyright header and clang-format didn't complain.
I'll give the style guide a read and get this fixed :).

**Review by AaronBallman — APPROVED** [review]
Thank you for the timing information! The yvals_core.h hack is actually a really clever way to handle this. I'm a bit worried that we don't have any test coverage for this file to begin with, so it's a bit hard to validate that the changes are correct, but we're early enough in a release cycle that I think we'll get feedback with plenty of time to address unintentional fallout.

The changes should come with a release note in `clang/docs/ReleaseNotes.rst` so that users know about the compile time performance improvements. Otherwise, this LGTM! I can land the changes on your behalf once there's a release note.

**Comment by MaxEW707** [comment]
Just want to say thanks for everyone taking the time to review the PR and providing feedback :).
Looking forward to using this in the next release of clang-cl.

> I'm a bit worried that we don't have any test coverage for this file to begin with, so it's a bit hard to validate that the changes are correct, but we're early enough in a release cycle that I think we'll get feedback with plenty of time to address unintentional fallout.

I'll look into adding unit tests next week in a separate PR. If not it will be closer towards the end of March due to personal things.
If any issues pop up feel free to @ me. I don't intend to commit and dash :).

> The changes should come with a release note in clang/docs/ReleaseNotes.rst so that users know about the compile time performance improvements.

Done. Feel free to reword the release notes as necessary. I believe I landed on something that is informative but succinct.

**Review by AaronBallman — APPROVED** [review]
LGTM, thank you!

**Review by CaseyCarter — APPROVED** [review]
Thanks for devising this clever way to improve MSVCSTL throughput!

**Comment by MaxEW707** [comment]
Friendly reminder that I require someone to commit on my behalf since I do not have write access :).

**Comment by bebuch** [comment]
@AaronBallman Can you merge this please? We are eagerly waiting for this to end up in a release! ;-)

**Comment by AaronBallman** [comment]
> @AaronBallman Can you merge this please? We are eagerly waiting for this to end up in a release! ;-)

Thank you for the ping, this fell off my radar!

**Pull request merged**
